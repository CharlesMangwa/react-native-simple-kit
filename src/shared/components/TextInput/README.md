# TextInput ✍️

A generic component for handling user text inputs.

### API

- autoFocus?: `boolean`,
- defaultValue?: `string`,
- keyboardType?: `string`,
- multiline?: `boolean`,
- onBlur?: `Function`,
- onChangeText?: `Function`,
- onFocus?: `Function`,
- onSubmitEditing?: `Function`,
- placeholder: `string`,
- placeholderTextColor: `string`,
- returnKeyType?: `string`,
- search?: `boolean`,
- secured?: `boolean`,
- style?: `StyleSheet`,
- value?: `any`,

### Example

```js
/* @flow */
import React from 'react'
import { StyleSheet, View } from 'react-native'
import TextInput from '@components/TextInput'

const MyComponent = (): React$Element<*> => (
  <View>
    /* Option A */
    <TextInput
      keyboardType="numeric"
      placeholder="Number"
      style={StyleSheet.flatten({ borderWidth: 1, borderColor: 'white' })
    />
    /* Option B */
    <TextInput secured placeholder="Password" />
  </View>
)

export default MyComponent
```

## Changelog

#### Saturday, June 23rd

- Refactoring for React 16.3+

#### Thursday, November 16th

- Initial release. We didn't track changes before this version.
