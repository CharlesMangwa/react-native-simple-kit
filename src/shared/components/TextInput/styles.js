/* @flow */

import { StyleSheet } from 'react-native'
import { rem } from '@helpers/responsive'
import { DEFAULT_BACKGROUND_COLOR, DEFAULT_TEXT_COLOR } from '@theme/colors'
import { AVENIR_REGULAR } from '@theme/fonts'

const styles: Object = StyleSheet.create({
  container: {
    ...AVENIR_REGULAR,
    alignSelf: 'center',
    justifyContent: 'flex-start',
    alignItems: 'center',
    width: '93%',
    height: rem(40),
    fontSize: rem(14),
    color: DEFAULT_TEXT_COLOR,
    backgroundColor: DEFAULT_BACKGROUND_COLOR,
    paddingLeft: rem(15),
  },
})

export default styles
