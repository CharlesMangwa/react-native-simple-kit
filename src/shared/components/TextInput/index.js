/* @flow */

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { TextInput as TextInputModule, StyleSheet } from 'react-native'

import { BRAND_COLOR_RED, DEFAULT_TEXT_COLOR } from '@theme/colors'
import type { DefaultProps, Props } from './types'
import styles from './styles'

class TextInput extends Component<Props> {
  ref: ?Component<*>

  static defaultProps: DefaultProps = {
    autoCapitalize: 'none',
    autoFocus: false,
    defaultValue: undefined,
    keyboardType: 'default',
    multiline: false,
    onBlur: undefined,
    onChangeText: undefined,
    onContentSizeChange: undefined,
    onFocus: undefined,
    onSubmitEditing: undefined,
    placeholderTextColor: undefined,
    returnKeyType: 'done',
    selectionColor: undefined,
    secured: false,
    style: StyleSheet.flatten(null),
    value: undefined,
  }

  render() {
    const {
      autoCapitalize,
      autoFocus,
      defaultValue,
      keyboardType,
      multiline,
      onBlur,
      onChangeText,
      onContentSizeChange,
      onFocus,
      onSubmitEditing,
      placeholder,
      placeholderTextColor,
      returnKeyType,
      selectionColor,
      secured,
      style,
      value,
    } = this.props
    return (
      <TextInputModule
        ref={c => (this.ref = c)}
        autoFocus={autoFocus}
        autoCapitalize={autoCapitalize || 'none'}
        blurOnSubmit={true}
        dataDetectorTypes="all"
        defaultValue={defaultValue}
        clearButtonMode="always"
        keyboardType={keyboardType || 'default'}
        multiline={multiline}
        onBlur={onBlur}
        onContentSizeChange={onContentSizeChange}
        onChangeText={onChangeText}
        onFocus={onFocus}
        onSubmitEditing={onSubmitEditing}
        placeholder={placeholder || ''}
        placeholderTextColor={placeholderTextColor || `${DEFAULT_TEXT_COLOR}80`}
        returnKeyType={returnKeyType || 'done'}
        secureTextEntry={secured}
        selectionColor={selectionColor || BRAND_COLOR_RED}
        style={[styles.container, style]}
        underlineColorAndroid="transparent"
        value={value}
      />
    )
  }
}

TextInput.propTypes = {
  autoCapitalize: PropTypes.oneOf(['none', 'sentences', 'words', 'characters']),
  autoFocus: PropTypes.bool,
  defaultValue: PropTypes.string,
  keyboardType: PropTypes.oneOf([
    'default',
    'email-address',
    'numeric',
    'phone-pad',
    'ascii-capable',
    'numbers-and-punctuation',
    'url',
    'number-pad',
    'name-phone-pad',
    'decimal-pad',
    'twitter',
    'web-search',
  ]),
  multiline: PropTypes.bool,
  onBlur: PropTypes.func,
  onChangeText: PropTypes.func,
  onContentSizeChange: PropTypes.func,
  onFocus: PropTypes.func,
  onSubmitEditing: PropTypes.func,
  placeholder: PropTypes.string.isRequired,
  placeholderTextColor: PropTypes.string,
  returnKeyType: PropTypes.oneOf([
    'done',
    'go',
    'next',
    'search',
    'send',
    'none',
    'previous',
    'default',
    'emergency-call',
    'google',
    'join',
    'route',
    'yahoo',
  ]),
  selectionColor: PropTypes.string,
  secured: PropTypes.bool,
  style: PropTypes.oneOfType([PropTypes.number, PropTypes.object]),
  value: PropTypes.string,
}

export default TextInput
