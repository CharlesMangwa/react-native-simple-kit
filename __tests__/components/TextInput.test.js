import React from 'react'
import TestRenderer from 'react-test-renderer'
import ShallowRenderer from 'react-test-renderer/shallow'
import TextInput from '@components/TextInput'

describe('Generic component: <TextInput />', () => {
  it('renders correctly', () => {
    const component = TestRenderer.create(
      <TextInput placeholder="I'm rendered \o/" />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
