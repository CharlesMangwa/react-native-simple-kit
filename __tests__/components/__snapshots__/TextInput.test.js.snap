// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic component: <TextInput /> renders correctly 1`] = `
<TextInput
  allowFontScaling={true}
  autoCapitalize="none"
  autoFocus={false}
  blurOnSubmit={true}
  clearButtonMode="always"
  dataDetectorTypes="all"
  keyboardType="default"
  multiline={false}
  placeholder="I'm rendered \\\\o/"
  placeholderTextColor="#33333380"
  returnKeyType="done"
  secureTextEntry={false}
  selectionColor="#E33258"
  style={
    Array [
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#fefefe",
        "color": "#333333",
        "fontFamily": "Avenir",
        "fontSize": 25.728326341809918,
        "fontWeight": "400",
        "height": 73.50950383374261,
        "justifyContent": "flex-start",
        "paddingLeft": 27.56606393765348,
        "width": "93%",
      },
      undefined,
    ]
  }
  underlineColorAndroid="transparent"
/>
`;
